File: frontend/lib/binaryDefaultSpec.ts
export type FieldSpec = {
  key: string;
  type: "string" | "int32" | "int64" | "float" | "uint8";
  len: number;
  fmt?: (value: any, precision: number) => string;
};

export type PacketSpec = {
  [fieldId: number]: FieldSpec;
};

export type PktSpecMap = {
  [pktType: number]: PacketSpec;
};

export const QUOTE = "quote";
export const QUOTE_2 = "quote2";
export const QUOTE_3 = "quote3";

export const PKT_TYPE: Record<number, string> = {
  49: "quote",
  50: "quote2",
  52: "quote3",
};

// Formatting functions
function commafmt(value: number, precision: number = 2): string {
  const v = Number(value).toFixed(precision);
  const parts = v.split(".");
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ",");
  return parts.join(".");
}

function datefmt(value: number): string {
  if (value == null) return String(value);
  const month = ["Jan", "Feb", "Mar", "Apr", "May", "June", "July", "Aug", "Sep", "Oct", "Nov", "Dec"];
  const date = new Date(value * 1000);
  const dd = date.getDate().toString().padStart(2, "0");
  const mm = month[date.getMonth()];
  const yyyy = date.getFullYear();
  const time = date.toLocaleTimeString("en-US", { hour12: true });
  return `${dd} ${mm} ${yyyy}, ${time}`;
}

export const DEFAULT_PKT_INFO: {
  PKT_SPEC: PktSpecMap;
} = {
  PKT_SPEC: {
    49: {
      65: { key: "symbol", type: "string", len: 20 },
      66: { key: "precision", type: "uint8", len: 1 },
      67: { key: "ltp", type: "float", len: 8, fmt: commafmt },
      68: { key: "open", type: "float", len: 8, fmt: commafmt },
      69: { key: "high", type: "float", len: 8, fmt: commafmt },
      70: { key: "low", type: "float", len: 8, fmt: commafmt },
      71: { key: "close", type: "float", len: 8, fmt: commafmt },
      72: { key: "chng", type: "float", len: 8, fmt: commafmt },
      73: { key: "chngPer", type: "float", len: 8, fmt: (v) => commafmt(v, 2) },
      74: { key: "atp", type: "float", len: 8, fmt: commafmt },
      75: { key: "yHigh", type: "float", len: 8, fmt: commafmt },
      76: { key: "yLow", type: "float", len: 8, fmt: commafmt },
      77: { key: "ltq", type: "int32", len: 4, fmt: (v) => commafmt(v, 0) },
      78: { key: "vol", type: "int32", len: 4, fmt: (v) => commafmt(v, 0) },
      79: { key: "ttv", type: "float", len: 8, fmt: commafmt },
      80: { key: "ucl", type: "float", len: 8, fmt: commafmt },
      81: { key: "lcl", type: "float", len: 8, fmt: commafmt },
      82: { key: "OI", type: "int32", len: 4, fmt: (v) => commafmt(v, 0) },
      83: { key: "OIChngPer", type: "float", len: 8, fmt: (v) => commafmt(v, 2) },
      84: { key: "ltt", type: "int32", len: 4, fmt: datefmt },
      87: { key: "bidprice", type: "float", len: 8, fmt: commafmt },
      90: { key: "askprice", type: "float", len: 8, fmt: commafmt },
    },
    50: {
      65: { key: "symbol", type: "string", len: 20 },
      66: { key: "precision", type: "uint8", len: 1 },
      85: { key: "totBuyQty", type: "int32", len: 4, fmt: (v) => commafmt(v, 0) },
      86: { key: "totSellQty", type: "int32", len: 4, fmt: (v) => commafmt(v, 0) },
      87: { key: "price", type: "float", len: 8, fmt: commafmt },
      88: { key: "qty", type: "int32", len: 4, fmt: (v) => commafmt(v, 0) },
      89: { key: "no", type: "int32", len: 4, fmt: (v) => commafmt(v, 0) },
      90: { key: "price", type: "float", len: 8, fmt: commafmt },
      91: { key: "qty", type: "int32", len: 4, fmt: (v) => commafmt(v, 0) },
      92: { key: "no", type: "int32", len: 4, fmt: (v) => commafmt(v, 0) },
      93: { key: "nDepth", type: "uint8", len: 1 },
    },
    52: {
      65: { key: "symbol", type: "string", len: 20 },
      99: { key: "iv", type: "float", len: 8, fmt: commafmt },
      100: { key: "atmiv", type: "float", len: 8, fmt: commafmt },
      101: { key: "delta", type: "float", len: 8, fmt: commafmt },
      102: { key: "theta", type: "float", len: 8, fmt: commafmt },
      103: { key: "vega", type: "float", len: 8, fmt: commafmt },
      104: { key: "gamma", type: "float", len: 8, fmt: commafmt },
    },
  },
};

File: frontend/binaryDefaultSpec.py
import re
from datetime import datetime

def commafmt(value, precision=2):
    v = str(round(float(value), 2))
    parts = v.split('.')
    parts[0] = re.sub(r'\B(?=(\d{3})+(?!\d))', ',', parts[0])
    return ".".join(parts)

def datefmt(value):
    if value is None:
        return value
    month = ["Jan", "Feb", "Mar", "Apr", "May", "June", "July", "Aug", "Sep", "Oct", "Nov", "Dec"]
    date_time = datetime.fromtimestamp(value)
    dd = int(date_time.strftime("%d"))
    mm = int(date_time.strftime("%m"))
    yyyy = int(date_time.strftime("%Y"))
    time = date_time.strftime("%H:%M:%S %p")
    if dd < 10:
        dd = '0' + str(dd)
    fullDate = str(dd) + ' ' + month[mm-1] + ' ' + str(yyyy) + ', ' + time
    return fullDate

QUOTE = "quote"
QUOTE_2 = "quote2"
QUOTE_3 = "quote3"

PKT_TYPE = {
    49: "quote",
    50: "quote2",
    52: "quote3"
}

DEFAULT_PKT_INFO = {
    "PKT_SPEC": {
        49: {
            65: {"type": "string", "key": "symbol", "len": 20},
            66: {"type": "uint8", "key": "precision", "len": 1},
            67: {"type": "float", "key": "ltp", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            68: {"type": "float", "key": "open", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            69: {"type": "float", "key": "high", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            70: {"type": "float", "key": "low", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            71: {"type": "float", "key": "close", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            72: {"type": "float", "key": "chng", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            73: {"type": "float", "key": "chngPer", "len": 8, "fmt": lambda v, p: commafmt(v, 2)},
            74: {"type": "float", "key": "atp", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            75: {"type": "float", "key": "yHigh", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            76: {"type": "float", "key": "yLow", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            77: {"type": "int32", "key": "ltq", "len": 4, "fmt": lambda v, p: commafmt(v, 0)},
            78: {"type": "int32", "key": "vol", "len": 4, "fmt": lambda v, p: commafmt(v, 0)},
            79: {"type": "float", "key": "ttv", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            80: {"type": "float", "key": "ucl", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            81: {"type": "float", "key": "lcl", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            82: {"type": "int32", "key": "OI", "len": 4, "fmt": lambda v, p: commafmt(v, 0)},
            83: {"type": "float", "key": "OIChngPer", "len": 8, "fmt": lambda v, p: commafmt(v, 2)},
            84: {"type": "int32", "key": "ltt", "len": 4, "fmt": lambda v, p: datefmt(v)},
            87: {"type": "float", "key": "bidprice", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            90: {"type": "float", "key": "askprice", "len": 8, "fmt": lambda v, p: commafmt(v, p)}
        },
        50: {
            65: {"type": "string", "key": "symbol", "len": 20},
            66: {"type": "uint8", "key": "precision", "len": 1},
            85: {"type": "int32", "key": "totBuyQty", "len": 4, "fmt": lambda v, p: commafmt(v, 0)},
            86: {"type": "int32", "key": "totSellQty", "len": 4, "fmt": lambda v, p: commafmt(v, 0)},
            87: {"type": "float", "key": "price", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            88: {"type": "int32", "key": "qty", "len": 4, "fmt": lambda v, p: commafmt(v, 0)},
            89: {"type": "int32", "key": "no", "len": 4, "fmt": lambda v, p: commafmt(v, 0)},
            90: {"type": "float", "key": "price", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            91: {"type": "int32", "key": "qty", "len": 4, "fmt": lambda v, p: commafmt(v, 0)},
            92: {"type": "int32", "key": "no", "len": 4, "fmt": lambda v, p: commafmt(v, 0)},
            93: {"type": "uint8", "key": "nDepth", "len": 1},
        },
        52: {
            65: {"type": "string", "key": "symbol", "len": 20},
            99: {"type": "float", "key": "iv", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            100: {"type": "float", "key": "atmiv", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            101: {"type": "float", "key": "delta", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            102: {"type": "float", "key": "theta", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            103: {"type": "float", "key": "vega", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
            104: {"type": "float", "key": "gamma", "len": 8, "fmt": lambda v, p: commafmt(v, p)},
        }
    },
    "BID_ASK_OBJ_LEN": 3
}

